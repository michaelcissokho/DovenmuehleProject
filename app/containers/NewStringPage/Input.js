import styled from 'styled-components';

const Input = styled.input`
  background-color: ${props => props.theme === 'dark' && 'black'};
  color: ${props => props.theme === 'dark' && 'white'};
`;

export default Input;
