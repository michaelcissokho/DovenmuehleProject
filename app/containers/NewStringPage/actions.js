import {
  SEND_STRING,
  ADD_STRING_ERROR,
  ADD_STRING_SUCCESS,
  CHANGE_STRING,
  RESET_STRING_FORM,
} from './constants';

export function changeString(string) {
  return {
    type: CHANGE_STRING,
    string,
  };
}

export function sendString(string) {
  return {
    type: SEND_STRING,
    string,
  };
}

export function addStringSuccess() {
  return {
    type: ADD_STRING_SUCCESS,
  };
}

export function addStringError() {
  return {
    type: ADD_STRING_ERROR,
  };
}

export function resetStringForm() {
  return {
    type: RESET_STRING_FORM,
  };
}
