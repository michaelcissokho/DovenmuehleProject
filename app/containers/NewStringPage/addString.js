import axios from 'axios';

const addString = string => {
  try {
    axios.post(`http://localhost:3000/api`, { newString: string });
    return new Promise(resolve => {
      resolve();
    });
  } catch (err) {
    return err;
  }
};

export default addString;
