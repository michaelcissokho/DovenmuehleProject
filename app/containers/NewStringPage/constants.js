export const SEND_STRING = 'SEND_STRING';
export const CHANGE_STRING = 'CHANGE_STRING';
export const ADD_STRING_SUCCESS = 'ADD_STRING_SUCCESS';
export const ADD_STRING_ERROR = 'ADD_STRING_ERROR';
export const RESET_STRING_FORM = 'RESET_STRING_FORM';
