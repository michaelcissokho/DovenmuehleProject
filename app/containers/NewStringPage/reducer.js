import {
  CHANGE_STRING,
  ADD_STRING_ERROR,
  ADD_STRING_SUCCESS,
  RESET_STRING_FORM,
} from './constants';

export const INITIAL_STATE = {
  string: '',
  success: false,
  error: false,
};

function addStringReducer(state = INITIAL_STATE, action) {
  switch (action.type) {
    case CHANGE_STRING:
      return { ...state, string: action.string };
    case ADD_STRING_SUCCESS:
      return { ...state, string: '', success: true, error: false };
    case ADD_STRING_ERROR:
      return { ...state, string: '', success: false, error: true };
    case RESET_STRING_FORM:
      return INITIAL_STATE;
    default:
      return state;
  }
}

export default addStringReducer;
