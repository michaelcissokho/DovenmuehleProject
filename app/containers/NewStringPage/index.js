import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import Input from './Input';
import addString from './addString';

const NewStringPage = () => {
  const INITIAL_VALUES = { string: '' };

  const [theme, setTheme] = useState('light');
  const [formData, setFormData] = useState(INITIAL_VALUES);

  // passes prop type of theme to input component
  const toggleTheme = () => {
    if (theme === 'light') {
      setTheme('dark');
    } else {
      setTheme('light');
    }
  };

  // handles change of characters to form item
  const handleChange = e => {
    setFormData({ string: e.target.value });
  };

  // handles submission of form
  const handleSubmit = e => {
    e.preventDefault();
    const { string } = formData;
    addString(string);
    // eslint-disable-next-line no-alert
    alert('String Has Been Added');
    setFormData(INITIAL_VALUES);
  };

  return (
    <div style={{ textAlign: 'center' }}>
      <h1>Add A New String:</h1>

      <div>
        <form onSubmit={handleSubmit}>
          <Input
            theme={theme}
            placeholder="New String"
            value={formData.string}
            onChange={handleChange}
            name="string"
            data-testid="inputBox"
          />
          <button type="submit"> Submit </button>
        </form>
        <h3>{theme.toUpperCase()} mode</h3>
        <button type="button" onClick={toggleTheme}>
          Toggle Theme
        </button>
      </div>

      <br />

      <Link to="/"> View All Strings </Link>
    </div>
  );
};

export default NewStringPage;
