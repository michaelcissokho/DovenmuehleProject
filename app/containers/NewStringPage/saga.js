import axios from 'axios';
import { call, put, takeLatest } from 'redux-saga/effects';
import { addStringError, addStringSuccess } from './actions';
import { SEND_STRING } from './constants';

function addString(string) {
  return axios.post(`http://localhost:3000/api/`, { newString: string });
}

function* handleAddString(action) {
  try {
    const { string } = action;
    yield call(addString, string);
    yield put(addStringSuccess());
  } catch (err) {
    yield put(addStringError());
  }
}

export default function* watchStringAddition() {
  yield takeLatest(SEND_STRING, handleAddString);
}
