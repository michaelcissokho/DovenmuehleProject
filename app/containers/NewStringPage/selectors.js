import { createSelector } from 'reselect';
import { INITIAL_STATE } from './reducer';

const stringForm = state => state.addStringForm || INITIAL_STATE;

const stringSelector = () =>
  createSelector(
    [stringForm],
    state => state.string,
  );

const addStringErrorSelector = () =>
  createSelector(
    [stringForm],
    state => state.error,
  );

const addStringSuccessSelector = () =>
  createSelector(
    [stringForm],
    state => state.success,
  );

export { stringSelector, addStringErrorSelector, addStringSuccessSelector };
