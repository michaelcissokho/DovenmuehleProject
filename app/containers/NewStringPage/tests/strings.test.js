/**
 * Test the NewStringPage container
 */

import React from 'react';
import { BrowserRouter } from 'react-router-dom';
// eslint-disable-next-line prettier/prettier
import {
  render,
  fireEvent,
} from 'react-testing-library';
import NewStringPage from '../index';

describe('<NewStringPage />', () => {
  it('should render and match the snapshot', () => {
    const newStringPage = render(
      <BrowserRouter>
        <NewStringPage />
      </BrowserRouter>,
    );
    expect(newStringPage).toMatchSnapshot();
  });

  describe('form submission', () => {
    describe('onSubmitString', () => {
      it('should start with Light theme and toggle to Dark Upon Click', () => {
        const { getByText, getByTestId } = render(
          <BrowserRouter>
            <NewStringPage />
          </BrowserRouter>,
        );

        const inputBox = getByTestId('inputBox');
        const classList = Object.values(inputBox.classList);
        expect(classList[1]).toBe(`fOXpjI`);

        const button = getByText('Toggle Theme');
        fireEvent.click(button);

        const newClassList = Object.values(inputBox.classList);

        // eslint-disable-next-line prettier/prettier
        expect(newClassList[1]).toBe(`hGYcCd`)
      });

      it('instantiate an alert from the window which means the addString function did not have an error', () => {
        const alertMock = jest.spyOn(window, 'alert').mockImplementation();

        const { getByText } = render(
          <BrowserRouter>
            <NewStringPage />
          </BrowserRouter>,
        );

        const submitButton = getByText('Submit');
        fireEvent.click(submitButton);

        expect(alertMock).toHaveBeenCalled();
      });
    });
  });
});
