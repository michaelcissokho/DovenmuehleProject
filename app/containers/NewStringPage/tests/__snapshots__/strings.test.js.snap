// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewStringPage /> should render and match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <h1>
          Add A New String:
        </h1>
        <div>
          <form>
            <input
              class="Input-sc-6bz4w7-0 fOXpjI"
              data-testid="inputBox"
              name="string"
              placeholder="New String"
              value=""
            />
            <button
              type="submit"
            >
               Submit 
            </button>
          </form>
          <h3>
            LIGHT
             mode
          </h3>
          <button
            type="button"
          >
            Toggle Theme
          </button>
        </div>
        <br />
        <a
          href="/"
        >
           View All Strings 
        </a>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h1>
        Add A New String:
      </h1>
      <div>
        <form>
          <input
            class="Input-sc-6bz4w7-0 fOXpjI"
            data-testid="inputBox"
            name="string"
            placeholder="New String"
            value=""
          />
          <button
            type="submit"
          >
             Submit 
          </button>
        </form>
        <h3>
          LIGHT
           mode
        </h3>
        <button
          type="button"
        >
          Toggle Theme
        </button>
      </div>
      <br />
      <a
        href="/"
      >
         View All Strings 
      </a>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllBySelectText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getBySelectText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllBySelectText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryBySelectText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
